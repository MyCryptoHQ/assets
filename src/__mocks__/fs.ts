import { basename } from 'path';

export const promises = {
  mkdir: jest.fn(),

  readFile: jest.fn(
    async (path: string, encoding: string): Promise<string> => {
      const file = basename(path);

      switch (file) {
        case 'eth.json':
          return `
            [
              {
                "address": "0xa74476443119A942dE498590Fe1f2454d7D4aC0d",
                "symbol": "GNT",
                "decimal": 18,
                "name": "Golem",
                "uuid": "59a6f871-34bb-5914-86cc-b5f1307bb3b9"
              },
              {
                "address": "0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2",
                "symbol": "MKR",
                "decimal": 18,
                "name": "MakerDAO",
                "uuid": "d3230312-cacb-52e6-a849-8bbf6e8142dd"
              }
            ]
          `;
        case 'assets.json':
          return `
          {
            "356a192b-7913-504c-9457-4d18c28d46e6": {
              "coinCapId": "ethereum",
              "coinGeckoId": "ethereum",
              "cryptoCompareId": "ETH",
              "cryptoCurrencyIconName": "eth",
              "invalid": "field"
            },
            "59a6f871-34bb-5914-86cc-b5f1307bb3b9": {
              "coinCapId": "golem-network-tokens",
              "coinGeckoId": "golem",
              "cryptoCompareId": "GNT",
              "cryptoCurrencyIconName": "gnt"
            },
            "d3230312-cacb-52e6-a849-8bbf6e8142dd": {
              "coinCapId": "maker",
              "coinGeckoId": "maker",
              "cryptoCompareId": "MKR",
              "cryptoCurrencyIconName": "mkr"
            }
          }
        `;
      }

      throw new Error('File not found');
    }
  ),

  writeFile: jest.fn()
};
