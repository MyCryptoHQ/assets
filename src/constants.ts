import { number, object, string } from '@hapi/joi';
import { resolve } from 'path';

export interface Asset {
  name: string;
  symbol: string;
  uuid: string;
  address?: string;
  decimal?: string;
}

export interface ParsedAsset {
  coinCapId?: string;
  coinGeckoId?: string;
  cryptoCompareId?: string;
  cryptoCurrencyIconName?: string;
  dexAgId?: string;
}

/**
 * Hardcoded base assets. The UUIDs are generated from the chain ID, e.g. `uuid('1')`.
 */
export const BASE_ASSETS: Asset[] = [
  {
    name: 'Ethereum',
    symbol: 'ETH',
    uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
  },
  {
    name: 'Ethereum Classic',
    symbol: 'ETC',
    uuid: '6c1e671f-9af5-546d-9c1a-52067bdf0e53'
  }
];

/**
 * Token asset validation schema.
 */
export const TOKEN_ASSET_SCHEMA = object({
  symbol: string().required(),
  address: string()
    .regex(/^0x[a-fA-F0-9]{40}$/)
    .required(),
  decimal: number()
    .min(0)
    .required(),
  name: string().required(),
  uuid: string().uuid()
});

/**
 * Parsed asset validation schema.
 */
export const ASSET_SCHEMA = object({
  coinCapId: string().optional(),
  coinGeckoId: string().optional(),
  cryptoCompareId: string().optional(),
  cryptoCurrencyIconName: string().optional(),
  dexAgId: string().optional()
});

export const OUTPUT_PATH = resolve(__dirname, '../assets/assets.json');
