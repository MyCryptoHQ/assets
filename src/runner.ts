import Listr from 'listr';
import { getAssets, getIds, getParsedAssets, writeToDisk } from './assets';
import { Asset, ParsedAsset } from './constants';
import { isEmpty } from './utils';

interface ApplicationContext {
  assets: Asset[];
  parsedAssets: Record<string, ParsedAsset>;
}

/**
 * Runs the application.
 */
export const run = async () => {
  const tasks = new Listr<ApplicationContext>([
    {
      title: 'Fetching assets',
      task: async context => {
        context.assets = await getAssets();
        context.parsedAssets = await getParsedAssets();
      }
    },
    {
      title: 'Matching assets with services',
      task: async context => {
        context.parsedAssets = await context.assets.reduce<Promise<Record<string, ParsedAsset>>>(
          async (promise, asset) => {
            const previous = await promise;
            const parsedAsset = previous[asset.uuid];
            const ids = await getIds(asset, parsedAsset);

            if (!isEmpty(ids)) {
              return {
                ...previous,
                [asset.uuid]: ids
              };
            }

            return {
              ...previous
            };
          },
          Promise.resolve(context.parsedAssets)
        );
      }
    },
    {
      title: 'Writing output to disk',
      task: context => writeToDisk(context.parsedAssets)
    }
  ]);

  return tasks.run();
};
