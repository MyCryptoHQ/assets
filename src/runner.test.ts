import { promises as fs } from 'fs';
import { run } from './runner';
import { Asset } from './constants';

jest.mock('fs');
jest.mock('./assets', () => ({
  getAssets: jest.fn(async () => [
    {
      name: 'Ethereum',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    },
    {
      name: 'Ethereum Classic',
      symbol: 'ETC',
      uuid: '6c1e671f-9af5-546d-9c1a-52067bdf0e53'
    }
  ]),

  getParsedAssets: jest.fn(async () => ({
    '356a192b-7913-504c-9457-4d18c28d46e6': {
      coinCapId: 'ethereum'
    }
  })),

  getIds: jest.fn(async (asset: Asset) => {
    switch (asset.uuid) {
      case '356a192b-7913-504c-9457-4d18c28d46e6':
        return {
          coinCapId: 'ethereum',
          coinGeckoId: 'ethereum',
          cryptoCompareId: 'ETH',
          cryptoCurrencyIconName: 'eth'
        };
      case '6c1e671f-9af5-546d-9c1a-52067bdf0e53':
        return {};
    }
  }),

  writeToDisk: jest.fn()
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('run', () => {
  it('runs the application', async () => {
    const context = await run();
    expect(context.assets).toMatchSnapshot();
    expect(context.parsedAssets).toMatchSnapshot();
  });
});
