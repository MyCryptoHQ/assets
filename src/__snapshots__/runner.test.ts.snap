// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run runs the application 1`] = `
Array [
  Object {
    "name": "Ethereum",
    "symbol": "ETH",
    "uuid": "356a192b-7913-504c-9457-4d18c28d46e6",
  },
  Object {
    "name": "Ethereum Classic",
    "symbol": "ETC",
    "uuid": "6c1e671f-9af5-546d-9c1a-52067bdf0e53",
  },
]
`;

exports[`run runs the application 2`] = `
Object {
  "356a192b-7913-504c-9457-4d18c28d46e6": Object {
    "coinCapId": "ethereum",
    "coinGeckoId": "ethereum",
    "cryptoCompareId": "ETH",
    "cryptoCurrencyIconName": "eth",
  },
}
`;
